sympy/integrals/rubi/Algebra_Integration_rules.py

```diff
@@ -0,0 +1,40 @@
+# File: SineIntegrationRules.py
+#
+# This file is a Python code generated from the publicly available rule-based
+# integrator (Rubi) system created by Albert Rich available at
+# http://www.apmaths.uwo.ca/~arich/. This file is under the modified BSD
+# #TODO: license (see the LICENSE file for more details).
+
+from sympy import simplify, log
+from sympy.core.symbol import Symbol, symbols
+from sympy.core.symbol import Wild
+
+def rubi_integrate(self, x):
+	"""
+	rubi_integrate decides which category of rule does the
+	given integrand (self) falls in and returns to that rule.
+	"""
+    if self.is_Mul:
+        if self.args[0].has(x) == False and len(self.args) == 2:
+            # condition to check if the first arg is a constant,
+            # returns in an is_Pow form so that it matches better.
+            return (self.args[0]*int_alg_1_1(self.args[1], x))
+    if self.is_Pow:
+        return (int_alg_1_1(self, x))
+
+def int_alg_1_1(self, x):
+    # rule int_alg_1_1 indicates Algebraic > Linear Product > (a + b*x)**m category
+    a = Wild('a', exclude = [x])
+    b = Wild('b', exclude = [x])
+    m = Wild('m', exclude = [x])
+    simplify(self)
+    type = self.match_between(((b*x)**m), (a + b*x)**m)
+    b = type[b]
+    m = type[m]
+    a = type[a]
```

Comment:
This will fail if it matches the first variant. You need some way to indicate that `a` should be 0 if the `(b*x)**m` pattern is used. 

Maybe you have other more advanced examples in mind, but it seems to me that it would be better to abstract this so that you just deal with the one pattern `(a + b*x)**m`, and it automatically splits that out into the different patterns:

```
(a + b*x)**m    (general case)
(b*x)**m        (a = 0)
(a + b)**m      (b = 1)
a + b*x         (m = 1)
```

In general, any time a certain values for a Wild variable causes the expression to be evaluated in a way that it doesn't look like the pattern, you have to check a different pattern. This could mean special constants (like adding 0 or multiplying by 1), or automatic simplification (`(a*b)**m -> a**m*b**m` when `a` and `b` are positive). 

The current pattern matcher tries to be smart about some of these cases (like multiplying by 1), but a purely structural pattern matcher won't know anything about the mathematics.